src/function.ts

```diff
@@ -608,6 +611,19 @@ export class Function<in Payload, Out, OutPayload extends Payload = Payload> ext
           onFailure: FunctionBase.normalizeAsyncDestination<OutPayload, Out>(onFailure),
         });
 
+        // Poison pill that forces Function synthesis to fail when the closure serialization has not completed.
+        // Closure synthesis runs async, but CDK does not normally support async.
+        // In order for the synthesis to complete successfully 
+        // 1. Use autoSynth `new App({ autoSynth: true })` or `new App()` with the CDK Cli (`cdk synth`)
+        // 2. Use `await asyncSynth(app)` exported from Functionless in place of `app.synth()`
+        // 3. Manually await on the closure serializer promises `await Promise.all(Function.promises)`
+        // https://github.com/functionless/functionless/issues/128
+        _resource.node.addValidation({
+          validate: () => this.resource.node.metadata.find(m => m.type === FUNCTION_CLOSURE_FLAG) 
+            ? [] 
+            : [formatErrorMessage(ErrorCodes.Function_Closure_Serialization_Incomplete)]
+        });
```

Comment:
When do these validations run?

Reply:
Immediately before synth, after app.synth is called.

I'm pretty excited about the find. Feel like there will be more uses for this...

The cool thing is that the validations do not short circuit, cdk shows all errors for all constructs. 

Which means we could do something like error on an assertion about least privilege at the last moment. 

() => table.neverWrite() // fails during synth if the lambda role contains write

Reply:
I was wondering how it plays with the promises? Wouldn't the validate have a chance of running before the promises have completed?

Reply:
It should run as part of synth, so if the promises have been awaited before synth, the validation will pass and if the promises have not been allowed to complete, they will fail.

There is a chance that the validation succeeds because the promises happen to complete before synth is called, in which case the state is correct.

1. serialize 
2. synth
   3. validate (fail if 1 has not completed - only possible by luck)

1. serialize
2. await
3. synth
  4. validate (pass)
  5. output

Reply:
What happens when using auto synth? The promises won't be awaited

Reply:
The node process waits for the promises before executing synth. I tested this with the cli. 

`cdk synth` works as long as `app.synth()` isn't called manually.